### PR TITLE
Add support for users updating the CSP's sources

### PR DESCRIPTION
**Pull request by Sigmonia**
#### Rationale
[Secure Issue 50544: UI-based support for customizing CSP-allowed servers](https://www.labkey.org/home/Developer/issues/Secure/issues-details.view?issueId=50544)
I added a new link in the Admin Console > Settings > Configuration column and reused the External Redirect UI, forms, and controller. 

#### Related PRs
* https://github.com/LabKey/testAutomation/pull/2105

#### Changes
- Generalized the ExternalRedirectAdminAction's JSPs, form, and methods
- Added a method to register the new source hosts with the CSP at startup and save

### CODE REVIEW

File: api/src/org/labkey/api/settings/AppPropsImpl.java [hunk 1]
```diff
@@ -625,7 +625,19 @@ public List<String> getExternalRedirectHosts()
...
+    @NotNull
+    public List<String> getExternalSourceHosts()
+    {
+        String urls = lookupStringValue(externalSourceHostURLs, "");
+        if (StringUtils.isNotBlank(urls))
+        {
+            return new ArrayList<>(Arrays.asList(urls.split(EXTERNAL_HOST_DELIMITER)));
         }
         return new ArrayList<>();
     }
```
Comment (labkey-adam) [inline api/src/org/labkey/api/settings/AppPropsImpl.java:643]: Could factor out a common method that takes the enum constant to eliminate duplicate code

File: api/src/org/labkey/api/settings/WriteableAppProps.java [hunk 2]
```diff
@@ -221,8 +221,18 @@ public void setExternalRedirectHosts(@NotNull List<String> externalRedirectHosts
...
+    public void setExternalSourceHosts(@NotNull List<String> externalSourceHosts)
+    {
+        StringBuilder urls = new StringBuilder();
+        for (String externalSourceURL : externalSourceHosts)
+        {
+            urls.append(externalSourceURL.trim()).append(EXTERNAL_HOST_DELIMITER);
+        }
+        storeStringValue(externalSourceHostURLs, urls.toString());
+    }
 }
```
Comment (labkey-adam) [inline api/src/org/labkey/api/settings/WriteableAppProps.java:233]: Factor out a common method here as well?

File: core/src/org/labkey/core/CoreModule.java [hunk 3]
```diff
@@ -1249,6 +1250,15 @@ public void moduleStartupComplete(ServletContext servletContext)
         UserManager.addUserListener(new EmailPreferenceUserListener());
     }
 
+    private void registerAllowedConnectionSources()
+    {
+        ContentSecurityPolicyFilter.registerAllowedConnectionSource(
+                AdminController.getExternalSourceHostsKey(),
+                AppProps.getInstance().getExternalSourceHosts().toArray(new String[0]));
+
+        LOG.debug("Registered [{}] as an allowed connection source", String.join(", ", AppProps.getInstance().getExternalSourceHosts()));
```
Comment (labkey-adam) [inline core/src/org/labkey/core/CoreModule.java]:
```suggestion
        LOG.debug("Registered [{}] as an allowed connection source", () -> String.join(", ", AppProps.getInstance().getExternalSourceHosts()));

```

Comment (labkey-adam) [reply core/src/org/labkey/core/CoreModule.java]: Totally optional, but this avoids doing the join work unless the level is `DEBUG`.

File: core/src/org/labkey/core/admin/AdminController.java [hunk 4]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
+        public void validateCommand(ExternalHostsForm target, Errors errors)
         {
         }
 
         @Override
-        public ModelAndView getView(ExternalRedirectForm form, boolean reshow, BindException errors)
+        public ModelAndView getView(ExternalHostsForm form, boolean reshow, BindException errors)
         {
-            form.setExistingRedirectHostList(AppProps.getInstance().getExternalRedirectHosts());
+            form.setExistingHostsList(form.getTypeEnum() == ExternalServerType.Source ? AppProps.getInstance().getExternalSourceHosts() : AppProps.getInstance().getExternalRedirectHosts());
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]: Cleaner (and more future expandable) would be to push the enum-specific details into each constant. So caller would be something like:
```
form.setExistingHostsList(form.getTypeEnum().getHostsList());
```

File: core/src/org/labkey/core/admin/AdminController.java [hunk 5]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
-        public boolean handlePost(ExternalRedirectForm form, BindException errors) throws Exception
+        public boolean handlePost(ExternalHostsForm form, BindException errors) throws Exception
+        {
+            WriteableAppProps appProps = AppProps.getWriteableInstance();
+            boolean update;
+
+            if (form.getTypeEnum() == ExternalServerType.Source)
+                update = handleHostPost(form, appProps::getExternalSourceHosts, appProps::setExternalSourceHosts, errors);
+            else
+                update = handleHostPost(form, appProps::getExternalRedirectHosts, appProps::setExternalRedirectHosts,  errors);
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]: Ideally, this code wouldn't know anything about the specific behaviors of the enum constants... just pass the enum constant into `handleHostPost()` and implement `getHosts()` and `setHosts()` on the enum.

File: core/src/org/labkey/core/admin/AdminController.java [hunk 6]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
+
+            if (update)
+            {
+                appProps.save(getUser());
+                if (form.getTypeEnum() == ExternalServerType.Source)
+                {
+                    // Refresh the CSP with new values.
+                    ContentSecurityPolicyFilter.unregisterAllowedConnectionSource(getExternalSourceHostsKey());
+                    ContentSecurityPolicyFilter.registerAllowedConnectionSource(getExternalSourceHostsKey(), appProps.getExternalSourceHosts().toArray(new String[0]));
+                }
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]: Similarly, implement `afterSave()` on the enum and call it unconditionally. Empty method in the case of Redirect.

File: core/src/org/labkey/core/admin/AdminController.java [hunk 7]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
+
+        private boolean handleHostPost(ExternalHostsForm form, Supplier<List<String>> getHosts, Consumer<List<String>> setHosts, BindException errors) throws Exception
         {
             //handle delete of existing external redirect host
             if (form.isDelete())
             {
                 String urlToDelete = form.getExistingExternalHost();
-                List<String> redirectHosts = AppProps.getInstance().getExternalRedirectHosts();
+                List<String> redirectHosts = getHosts.get();
                 for (String externalRedirectHost : redirectHosts)
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]:
```suggestion
                for (String externalHost : externalHosts)

```

File: core/src/org/labkey/core/admin/AdminController.java [hunk 8]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
+        }
+
+        private boolean handleHostPost(ExternalHostsForm form, Supplier<List<String>> getHosts, Consumer<List<String>> setHosts, BindException errors) throws Exception
         {
             //handle delete of existing external redirect host
             if (form.isDelete())
             {
                 String urlToDelete = form.getExistingExternalHost();
-                List<String> redirectHosts = AppProps.getInstance().getExternalRedirectHosts();
+                List<String> redirectHosts = getHosts.get();
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]:
```suggestion
                List<String> externalHosts = getHosts.get();

```

File: core/src/org/labkey/core/admin/AdminController.java [hunk 9]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
                 for (String externalRedirectHost : redirectHosts)
                 {
                     if (null != urlToDelete && urlToDelete.trim().equalsIgnoreCase(externalRedirectHost.trim()))
                     {
                         redirectHosts.remove(externalRedirectHost);
-                        WriteableAppProps appProps = AppProps.getWriteableInstance();
-                        appProps.setExternalRedirectHosts(redirectHosts);
-                        appProps.save(getUser());
+                        setHosts.accept(redirectHosts);
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]:
```suggestion
                        setHosts.accept(externalHosts);

```

File: core/src/org/labkey/core/admin/AdminController.java [hunk 10]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
                         break;
                     }
                 }
             }
             //handle updates - clicking on Save button under Existing will save the updated urls
             else if (form.isSaveAll())
             {
-                List<String> redirectHosts = form.getExistingRedirectHostList(); //get hosts from the form, this includes updated hosts
-                if (null != redirectHosts && redirectHosts.size() > 0)
+                List<String> redirectHosts = form.getExistingHostList(); //get hosts from the form, this includes updated hosts
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]: Rename... `externalHosts`?

File: core/src/org/labkey/core/admin/AdminController.java [hunk 11]
```diff
@@ -10699,13 +10727,11 @@ else if (form.isSaveNew())
                 }
                 else if (StringUtils.isNotEmpty(newExternalRedirectHost))
                 {
-                    List<String> existingRedirectHosts = AppProps.getInstance().getExternalRedirectHosts();
+                    List<String> existingRedirectHosts = getHosts.get();
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]: Rename?

File: core/src/org/labkey/core/admin/AdminController.java [hunk 12]
```diff
@@ -10690,7 +10718,7 @@ else if (form.isSaveAll())
             //save new external redirect host
             else if (form.isSaveNew())
             {
-                String newExternalRedirectHost = StringUtils.trimToEmpty(form.getNewExternalRedirectHost());
+                String newExternalRedirectHost = StringUtils.trimToEmpty(form.getNewExternalHost());
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]: Rename

File: core/src/org/labkey/core/admin/AdminController.java [hunk 13]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
         @Override
-        public ModelAndView getView(ExternalRedirectForm form, boolean reshow, BindException errors)
+        public ModelAndView getView(ExternalHostsForm form, boolean reshow, BindException errors)
         {
-            form.setExistingRedirectHostList(AppProps.getInstance().getExternalRedirectHosts());
+            form.setExistingHostsList(form.getTypeEnum() == ExternalServerType.Source ? AppProps.getInstance().getExternalSourceHosts() : AppProps.getInstance().getExternalRedirectHosts());
 
-            JspView<ExternalRedirectForm> newView = new JspView<>("/org/labkey/core/admin/addNewExternalRedirectHost.jsp", form, errors);
-            newView.setTitle("Register New External Redirect Host");
+            JspView<ExternalHostsForm> newView = new JspView<>("/org/labkey/core/admin/addNewExternalRedirectHost.jsp", form, errors);
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]: Should probably change the JSP names

File: core/src/org/labkey/core/admin/AdminController.java [hunk 14]
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
...
+            newView.setTitle(String.format("Register New External %1$s Host", form.getTypeEnum().name()));
             newView.setFrame(WebPartView.FrameType.PORTAL);
-            JspView<ExternalRedirectForm> existingView = new JspView<>("/org/labkey/core/admin/existingExternalRedirectHosts.jsp", form, errors);
-            existingView.setTitle("Existing External Redirect Hosts");
+            JspView<ExternalHostsForm> existingView = new JspView<>("/org/labkey/core/admin/existingExternalRedirectHosts.jsp", form, errors);
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java]: Should probably change the JSP names

File: core/src/org/labkey/core/admin/AdminController.java [hunk 15]
```diff
@@ -10747,9 +10773,9 @@ private boolean isDuplicate(List<String> existingURLs, String newExternalRedirec
...
 
         @Override
-        public URLHelper getSuccessURL(ExternalRedirectForm form)
+        public URLHelper getSuccessURL(ExternalHostsForm form)
         {
-            return new ActionURL(ExternalRedirectAdminAction.class, getContainer());
+            return new ActionURL(ExternalRedirectAdminAction.class, getContainer()).addParameter("type", form.getTypeEnum().name());
         }
 
         @Override
```
Comment (labkey-adam) [inline core/src/org/labkey/core/admin/AdminController.java:10819]: Need to update `addNavTrail()` as well. A help topic needs to be defined on the enum and used here. And the title needs to change based on the enum.

**Review by labkey-adam — CHANGES_REQUESTED** [review]

File: api/src/org/labkey/api/settings/WriteableAppProps.java [hunk 16]
```diff
@@ -216,13 +216,18 @@ public void setXFrameOption(String option)
...
+    }
+
+    public void setExternalSourceHosts(@NotNull Collection<String> externalSourceHosts)
+    {
+        setExternalHosts(externalSourceHostURLs, externalSourceHosts);
+    }
+
+    private void setExternalHosts(RandomStartupProperties propName, @NotNull Collection<String> externalSourceHosts)
+    {
+        storeStringValue(propName, String.join(EXTERNAL_HOST_DELIMITER, externalSourceHosts));
```
Comment (labkey-adam) [inline api/src/org/labkey/api/settings/WriteableAppProps.java:231]: Previous code did a `trim()` on each url... no longer needed?

**Review by labkey-adam — APPROVED** [review]
One question about trimming hosts before storing

File: api/src/org/labkey/api/settings/WriteableAppProps.java (hunk 16 above)
Comment (Sigmonia) [reply api/src/org/labkey/api/settings/WriteableAppProps.java:231]: I don't believe so. I moved the trim to the validation & set-addition step. I guess that opens a potential hole if we add new callers, but I think "trimming" fits better as part of the validation/sanitation process.

Comment (labkey-adam) [reply api/src/org/labkey/api/settings/WriteableAppProps.java:231]: Cool, I missed that

**Pull request merged**